src/mumble/ServerHandler.cpp

```diff
@@ -392,6 +393,11 @@ void ServerHandler::sendPing() {
 	if (!connection)
 		return;
 
+	if (iInFlightTCPPings >= g.s.iMaxInFlightTCPPings) {
+		serverConnectionClosed(QAbstractSocket::UnknownSocketError, tr("Server is not responding to TCP pings"));
```

Comment:
I think we should remove the `disconnect` method and custom even `exit` logic.
For the one case it is used (connected to server has no certificate) we don’t have to close Mumble, but just the Connection.

The `Connection` signal `connectionClosed` already has a `reason` parameter, which it passes to/is connected to the slot `serverConnectionClosed`.
We could thus add the (optional) reason paramter to the `disconnectSocket` method for our use-case - disconnecting the (no-longer-existant) connection with a reason.
I don’t get the check in `disconnectSocket` for emitting the `connectionClosed` though. Alternatively we could add a new method `Connection::disconnect(reason)` (or similar).

---

My thought process/details. Nvm if the above is enough:

`serverConnectionClosed` is a slot. Thus, it should be called (only) as a reaction to signals/events.

In `serverConnectionTimeoutOnConnect` it is called as a followup event/slot.

Just from the wording, without having checked the code, adding a parameter to the disconnect method feels better. `disconnect` is a verb, an "action method" rather than a slot/event.

Looking at the code, maybe we want to call disconnect/abort on `Connection` instead? Which would emit `connectionClosed`, which is connected to `serverConnectionClosed`.
Of course, we would have pass the reason through all of that. The `connectionClosed` signal already has/uses that reason parameter.

The `ServerHandler::disconnect()` method is only called in `ServerHandler::serverConnectionConnected()`, with a valid `Connection` object in place. So I guess the `ServerHandler::disconnect()` should be removed. We can close the connection on the `Connection` object - should be its responsibility.